### ISSUE TITLE
round trip bug + type inference bug

### ISSUE THREAD

**Issue by stew** [unisonweb/unison#2712]
Below is a transcript which show's a round trip problem but I kinda think that the reason that this is a round trip bug is that it's actually exposing an problem with ability inference. When the pretty printer prints `susan: Nat ->{} Nat -> ()` it prints it without the null ability marker: `susan: Nat -> Nat -> ()` which causes the error:
```
  The expression in red
  
                needs the abilities: {𝕖6}
    but was assumed to only require: {𝕖26, 𝕖5}
  
  This is likely a result of using an un-annotated function as an argument with concrete abilities. Try adding an annotation to the function definition whose body is red.
  
      7 |   mapWithKey susan pam
```

==TRANSCRIPT==

    ```unison
    unique type Map k v = Tip | Bin Nat k v (Map k v) (Map k v)
    
    mapWithKey : (k ->{e} a ->{e} b) -> Map k a ->{e} Map k b
    mapWithKey f m = Tip
    ```
    
    ```ucm
    .> add
    ```
    
    ```unison
    
    naiomi = 
      susan: Nat ->{} Nat -> ()
      susan a b = ()
      
      pam: Map Nat Nat
      pam = Tip
      
      mapWithKey susan pam
      
    ```
    
    ```ucm
    .> add
    .> edit naiomi
    .> undo
    ```
    
    ```ucm
    .> load scratch.u
    .> add
    ```

**Comment by atacratic** [comment]
Related: #738 #717 #703 - and see also https://gist.github.com/atacratic/7a91901d5535391910a2d34a2636a93c#ability-inference-and-generalization.  

If the limitation in inference shown in this issue is just how it has to be, then I think that's a good case for remembering and faithfully pretty-printing the null ability list.

**Comment by pchiusano** [comment]
Just to clarify, when `naiomi` is pretty-printed, the `{}` on the annotation of `susan` is elided by the pretty-printer. This causes round trip bugs.

**Comment by pchiusano** [comment]
Since there's a decent workaround (just rename the namespace to not use `_`) this is lower priority.

**Comment by aryairani** [comment]
> Since there's a decent workaround (just rename the namespace to not use `_`) this is lower priority.

@pchiusano Was this meant for this ticket? or for #2822

**Comment by aryairani** [comment]
Hopefully will fix these definitions from [website](https://github.com/unisonweb/website:#8agaqj52kh):
* [ ] fundamentals.controlFlow.exceptionHandling.either

**Comment by pchiusano** [comment]
Fixed by #2895
